src/MongoDB.Driver.Core/Core/Authentication/External/AwsAuthenticationCredentialsProvider.cs

```diff
@@ -0,0 +1,202 @@
+        private readonly string _sessionToken;
+
+        public AwsCredentials(string accessKeyId, SecureString secretAccessKey, string sessionToken)
+        {
+            _accessKeyId = Ensure.IsNotNull(accessKeyId, nameof(accessKeyId));
+            _secretAccessKey = Ensure.IsNotNull(secretAccessKey, nameof(secretAccessKey));
+            _sessionToken = sessionToken; // can be null
+        }
+
+        public string AccessKeyId => _accessKeyId;
+        public SecureString SecretAccessKey => _secretAccessKey;
+        public string SessionToken => _sessionToken;
+
+        public BsonDocument GetKmsCredentials()
+            => new BsonDocument
+            {
+                { "accessKeyId", _accessKeyId },
+                { "secretAccessKey", SecureStringHelper.ToInsecureString(_secretAccessKey) },
+                { "sessionToken", _sessionToken, _sessionToken != null }
+            };
+    }
+
+    internal class AwsAuthenticationCredentialsProvider : IExternalAuthenticationCredentialsProvider<AwsCredentials>
+    {
+        private readonly AwsHttpClientHelper _awsHttpClientHelper;
+        private readonly HttpClientHelper _httpClientHelper;
+
+        public AwsAuthenticationCredentialsProvider(HttpClientHelper httpClientHelper)
+        {
+            _httpClientHelper = Ensure.IsNotNull(httpClientHelper, nameof(httpClientHelper));
+            _awsHttpClientHelper = new AwsHttpClientHelper(_httpClientHelper);
+        }
+
+        public AwsCredentials CreateCredentialsFromExternal(CancellationToken cancellationToken) =>
+            CreateCredentialsFromExternalAsync(cancellationToken).GetAwaiter().GetResult();
+        public async Task<AwsCredentials> CreateCredentialsFromExternalAsync(CancellationToken cancellationToken)
+        {
+            return CreateAwsCredentialsFromEnvironmentVariables() ??
+                (await CreateAwsCredentialsFromEcsResponseAsync(cancellationToken).ConfigureAwait(false)) ??
+                (await CreateAwsCredentialsFromEc2ResponseAsync(cancellationToken).ConfigureAwait(false)) ??
+                throw new InvalidOperationException($"Unable to find credentials for AWS authentication.");
+        }
+
+        private AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
+        {
+            var accessKeyId = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
+            var secretAccessKey = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
+            var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");
+
+            if (accessKeyId == null && secretAccessKey == null && sessionToken == null)
+            {
+                return null;
+            }
+            if (secretAccessKey != null && accessKeyId == null)
+            {
+                throw new InvalidOperationException($"When using AWS authentication if a secret access key is provided via environment variables then an access key ID must be provided also.");
```

Comment:
initially I provided a `subject` method argument to be able to have 100% the same message as before:
`...When using MONGODB-AWS authentication...`  instead the current version (without `MONGODB-` part), but given that a spec doesn't require particular exception, we can be ok with slight message change for simplicity?